src/v/datalake/translation/state_machine.cc

```diff
@@ -174,13 +174,29 @@ model::offset translation_stm::max_collectible_offset() {
     if (!_raft->log_config().iceberg_enabled()) {
         return model::offset::max();
     }
-    // if offset is not initialized, do not attempt translation.
+    // If offset is not initialized (we've never translated anything), avoid
+    // removing data. This means that when we haven't yet translated anything,
+    // local data is pinned until we start translating. Since the first
+    // translation starts at the beginning of the local log, this helps ensure:
+    // - on a new partition, we translate from 0 without reading from cloud
+    // - on a recovery topic partition, we don't backfill historical data
     if (_highest_translated_offset == kafka::offset{}) {
-        return model::offset{};
+        return model::offset::min();
     }
+    // If we have translated data before, no need to pin local data because
+    // translators don't need to read local data.
+    return model::offset::max();
+}
 
-    return highest_log_offset_below_next(
-      _raft->log(), _highest_translated_offset);
+std::optional<kafka::offset> translation_stm::lowest_pinned_offset() const {
```

Comment:
Does it have to be Kafka offset ?, we are mixing model and Kafka offsets in the same interface

Reply:
For posterity, we chatted about this a bit over a call and while I agree that it's a bit rough to be mixing Kafka and model offsets in the STM manager,
1. we do need this to be Kafka offsets, because the translation STM is unaware of model offsets, especially when reading data from tiered storage
2. _something_ needs to know about both Kafka offsets and model offsets, whether it's the translator STM, or cloud storage readers, or the STM manager. The STM manager just so happens to be the lowest hanging fruit.

I'm open to moving this around. The benefit of the current approach is that there are roughly no on-disk changes, so it seems like it should be easy to move concepts around.